src/cli_encryption_key/generate.js

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { safeDump } from 'js-yaml';
+import { appendFileSync } from 'fs';
+import { isEmpty } from 'lodash';
+import { join } from 'path';
+
+import { getConfigDirectory } from '@kbn/utils';
+
+import { confirm } from '../cli_keystore/utils';
+import { Logger } from '../cli_plugin/lib/logger';
+
+export async function generate(encryptionConfig, command) {
+  const logger = new Logger();
+  const keys = encryptionConfig.generate({ force: command.force });
+  if (!command.quiet) {
+    if (command.force) {
+      logger.log('Regenerating all encryption keys.  Add these to kibana.yml:');
+    } else {
+      logger.log('Generating missing encryption keys.  Add these to kibana.yml:');
+    }
+  }
+  if (isEmpty(keys)) {
+    logger.log('No keys to write.  Use the --force flag to generate new keys.');
+  } else {
+    logger.log(safeDump(keys));
+  }
+
+  if (command.interactive) {
+    const write = await confirm('Write to kibana.yml?');
+    if (write) {
+      const kibanaYML = join(getConfigDirectory(), 'kibana.yml');
+      appendFileSync(kibanaYML, safeDump(keys));
```

Comment:
I'm up for adding a more nuanced interactive mode.  Any thoughts @legrego @kobelb @tylersmalley?

For the conflict between interactive and force - the sticking point was key rotation.  We'll want to be careful about any removal, e.g. dropping a reporting encryption key could cause reports to be lost.

Reply:
>  it seems interactive with force can lead to an invalid config because of duplicated entries.

Yeah we definitely don't want to leave our users with invalid `kibana.yml` files. 


> I'm up for adding a more nuanced interactive mode.

This makes sense to me, and would be consistent with some of our other CLI utilities, like `elasticsearch-setup-passwords`.

> For the conflict between interactive and force - the sticking point was key rotation. We'll want to be careful about any removal, e.g. dropping a reporting encryption key could cause reports to be lost.

I don't fully understand this point. Isn't `--force` designed to replace the existing keys? If so, then I think replacing these keys with `--interactive` would make sense. We could have a more strongly worded confirmation message when both `--interactive` and `--force` are provided, but otherwise I'd think that replacing the existing keys is expected behavior.

Reply:
Works for me.  I'll make another iteration on the interactive portions.

> I don't fully understand this point. Isn't --force designed to replace the existing keys? If so, then I think replacing these keys with --interactive would make sense. We could have a more strongly worded confirmation message when both --interactive and --force are provided, but otherwise I'd think that replacing the existing keys is expected behavior.

My original planning was that it would not infrequently cause someone to back out to rotate.  Opposed to warning followed by a failure to start would force an acknowledgement via manual triage.  Mostly I didn't want to delete anything that  that could be sensitive, e.g. technically valid yaml for an encryption key could span two lines.

Reply:
I pushed 6240c79 for UX feedback.  It includes prompts for each setting.  

Any thoughts on the force flag?  I'm tempted to remove it altogether, key removal will require us to manage yaml parsing in a way that doesn't interfere with comments.  I can edit the message to tell the user to rotate keys that are set if there isn't anything to generate

Reply:
> I pushed 6240c79 for UX feedback. It includes prompts for each setting.
> Any thoughts on the force flag? I'm tempted to remove it altogether, key removal will require us to manage yaml parsing in a way that doesn't interfere with comments. I can edit the message to tell the user to rotate keys that are set if there isn't anything to generate

Yeah, I'm also worried about editing user's `kibana.yml`, but how about generating a sample `kibana.yaml` instead, something like [this one](https://github.com/elastic/elasticsearch/blob/39a5d755c99f1ef921fb2621b851e5f0a05a9da0/x-pack/plugin/security/cli/src/main/resources/org/elasticsearch/xpack/security/cli/certutil-http/kibana-sample.yml) generated by `elasticsearch-certutil`?

In the latest version of the PR we have this:
```
Generating settings for:
xpack.encryptedSavedObjects.encryptionKey
xpack.reporting.encryptionKey
xpack.security.encryptionKey

Set xpack.encryptedSavedObjects.encryptionKey? [y/N] 
Set xpack.reporting.encryptionKey? [y/N] 
Set xpack.security.encryptionKey? [y/N] 
Write 0 setting to kibana.yml? [y/N]
```
Maybe we can borrow some ideas from the `elasticsearch-certutil` here as well and have something similar to this (text is up to review by Docs Team and/or native speakers :slightly_smiling_face: )?
```
## Kibana Encryption Key Generation Utility

The 'generate' command guides you through the process of generating encryption
keys for the ........

This tool will ask you a number of questions in order to generate the right
set of keys for your needs.

Generate xpack.encryptedSavedObjects.encryptionKey? [y/N] y
Generate xpack.reporting.encryptionKey? [y/N] N
Generate xpack.security.encryptionKey? [y/N] y

The following encryption keys were generated:

xpack.encryptedSavedObjects.encryptionKey
xpack.security.encryptionKey

Save generated keys to a sample Kibana configuration file? Otherwise settings will be outputed here [y/N] y

What filename should be used for the sample Kibana config file? [/home/azasypkin/..../sample-kibana.yml]

----OR-----

xpack.encryptedSavedObjects.encryptionKey: 487f11e261c30cdca21a815521fe08cc
xpack.security.encryptionKey: 7b3b357dc84d4689bcf554ed1e599aed
```

Explanation of the key rotation in case `xpack.encryptedSavedObjects.encryptionKey` is generated can either go to a sample Kibana yml as a comment section or we can output it to STDOUT (with a link to https://www.elastic.co/guide/en/kibana/master/xpack-security-secure-saved-objects.html#encryption-key-rotation maybe)

What do you all think?

Reply:
Thanks, I like that idea.  I'll get that and the tests updated.

Reply:
Here's the latest:

auto
```
## Kibana Encryption Key Generation Utility

The 'generate' command guides you through the process of setting encryption keys for:
xpack.encryptedSavedObjects.encryptionKey
    Used to encrypt stored objects such as dashboards and visualizations
    https://www.elastic.co/guide/en/kibana/current/xpack-security-secure-saved-objects.html#xpack-security-secure-saved-objects
xpack.reporting.encryptionKey
    Used to encrypt saved reports
    https://www.elastic.co/guide/en/kibana/current/security-settings-kb.html#security-session-and-cookie-settings
xpack.security.encryptionKey
    Used to encrypt session information
    https://www.elastic.co/guide/en/kibana/current/security-settings-kb.html#security-session-and-cookie-settings

Already defined settings are ignored and can be regenerated using the --force flag.  Check the documentation links for instructions on how to rotate encryption keys.
Definitions should be set in the kibana.yml used configure Kibana.

Settings:
xpack.encryptedSavedObjects.encryptionKey: 37538fff48dd0929c6c736caaec3645f
xpack.reporting.encryptionKey: 6ff208df97f25c14b943e7f69b00141b
xpack.security.encryptionKey: cb38085389e3f12b29423e675f05ed11
```

interactive:
```
## Kibana Encryption Key Generation Utility

The 'generate' command guides you through the process of setting encryption keys for:
xpack.encryptedSavedObjects.encryptionKey
    Used to encrypt stored objects such as dashboards and visualizations
    https://www.elastic.co/guide/en/kibana/current/xpack-security-secure-saved-objects.html#xpack-security-secure-saved-objects
xpack.reporting.encryptionKey
    Used to encrypt saved reports
    https://www.elastic.co/guide/en/kibana/current/security-settings-kb.html#security-session-and-cookie-settings
xpack.security.encryptionKey
    Used to encrypt session information
    https://www.elastic.co/guide/en/kibana/current/security-settings-kb.html#security-session-and-cookie-settings

Already defined settings are ignored and can be regenerated using the --force flag.  Check the documentation links for instructions o
n how to rotate encryption keys.
Definitions should be set in the kibana.yml used configure Kibana.

This tool will ask you a number of questions in order to generate the right set of keys for your needs.

Set xpack.encryptedSavedObjects.encryptionKey? [y/N] y
Set xpack.reporting.encryptionKey? [y/N] n
Set xpack.security.encryptionKey? [y/N] y

The following keys were generated:
xpack.encryptedSavedObjects.encryptionKey
xpack.security.encryptionKey

Save generated keys to a sample Kibana configuration file? [y/N] n

Settings:
xpack.encryptedSavedObjects.encryptionKey: a4c722998ded91ab56e73bafdf743d14
xpack.security.encryptionKey: d87732b078669dd714acbb3af8fb8af6
```

I'm wondering if the command `generate` should be replaced `auto` and `interactive` similar to the elasticsearch util.  The quiet flag could be moved to auto to avoid any confusion.

Reply:
@gchaps if you have a chance, do you have any thoughts on the readability of the code blocks above?